src/ray/gcs/gcs_server/gcs_job_manager.cc

```diff
@@ -146,13 +147,66 @@ void GcsJobManager::HandleGetAllJobInfo(rpc::GetAllJobInfoRequest request,
                                         rpc::GetAllJobInfoReply *reply,
                                         rpc::SendReplyCallback send_reply_callback) {
   RAY_LOG(INFO) << "Getting all job info.";
-  auto on_done = [reply, send_reply_callback](
+  auto on_done = [this, reply, send_reply_callback](
                      const absl::flat_hash_map<JobID, JobTableData> &result) {
     for (auto &data : result) {
       reply->add_job_info_list()->CopyFrom(data.second);
     }
-    RAY_LOG(INFO) << "Finished getting all job info.";
-    GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::OK());
+
+    std::shared_ptr<std::atomic<int>> num_processed_jobs =
+        std::make_shared<std::atomic<int>>(0);
+
+    // For each job with a submission id, we query the internal KV store to get the Ray
+    // Job API JobInfo, and send the gRPC reply in the KV callback. If there are *no* jobs
+    // with a submission id, we must send the gRPC reply in the current function instead.
+    // Use this flag to track it.
+    bool any_jobs_with_submission_id = false;
+
+    for (int i = 0; i < reply->job_info_list_size(); i++) {
+      const auto &metadata = reply->job_info_list(i).config().metadata();
+      if (metadata.find("job_submission_id") == metadata.end()) {
+        num_processed_jobs->fetch_add(1);
+        continue;
+      }
+
+      any_jobs_with_submission_id = true;
+      const auto &job_submission_id = metadata.at("job_submission_id");
+
+      auto kv_get_callback =
+          [reply, send_reply_callback, num_processed_jobs, i, &job_submission_id](
+              std::optional<std::string> job_info_json) {
+            if (job_info_json) {
+              // Parse the JSON into a JobsAPIInfo proto.
+              rpc::JobsAPIInfo jobs_api_info;
+              RAY_CHECK(google::protobuf::util::JsonStringToMessage(*job_info_json,
```

Comment:
There has been past experience where this call will likely dominate the runtime - what's the scale of number of jobs with submission id? And if there's any requirement on the latency of this `GetALLJobInfo`?

Reply:
Not sure, but we should assume it's on the order of the total number of jobs across the lifetime of the cluster (including those without submission_id).  Do we have an estimate of this value? (From observability maybe?)

Not sure about the latency requirement of the endpoint, @ericl any thoughts about this?